fraud/bad_encoding.go

```diff
@@ -0,0 +1,160 @@
+package fraud
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"github.com/tendermint/tendermint/pkg/consts"
+	"github.com/tendermint/tendermint/pkg/wrapper"
+
+	"github.com/celestiaorg/rsmt2d"
+
+	pb "github.com/celestiaorg/celestia-node/fraud/pb"
+	"github.com/celestiaorg/celestia-node/ipld"
+	ipld_pb "github.com/celestiaorg/celestia-node/ipld/pb"
+	"github.com/celestiaorg/celestia-node/service/header"
+)
+
+type BadEncodingProof struct {
+	BlockHeight uint64
+	// ShareWithProof contains all shares from row or col.
+	// Shares that did not pass verification in rmst2d will be nil.
+	// For non-nil shares MerkleProofs are computed.
+	Shares []*ipld.NamespacedShareWithProof
+	// Index represents the row/col index where ErrByzantineRow/ErrByzantineColl occurred
+	Index uint8
+	// isRow shows that verification failed on row
+	isRow bool
+}
+
+// CreateBadEncodingProof creates a new Bad Encoding Fraud Proof that should be propagated through network
+// The fraud proof will contain shares that did not pass verification and their relevant Merkle proofs.
+func CreateBadEncodingProof(
+	height uint64,
+	shares *ipld.ErrByzantine,
+) Proof {
+
+	return &BadEncodingProof{
+		BlockHeight: height,
+		Shares:      shares.Shares,
+		isRow:       shares.IsRow,
+		Index:       shares.Index,
+	}
+}
+
+// Type returns type of fraud proof
+func (p *BadEncodingProof) Type() ProofType {
+	return BadEncoding
+}
+
+// Height returns block height
+func (p *BadEncodingProof) Height() uint64 {
+	return p.BlockHeight
+}
+
+// MarshalBinary converts BadEncodingProof to binary
+func (p *BadEncodingProof) MarshalBinary() ([]byte, error) {
+	shares := make([]*ipld_pb.Share, 0, len(p.Shares))
+	for _, share := range p.Shares {
+		shares = append(shares, share.ShareWithProofToProto())
+	}
+
+	badEncodingFraudProof := pb.BadEncoding{
+		Height: p.BlockHeight,
+		Shares: shares,
+		Index:  uint32(p.Index),
+		IsRow:  p.isRow,
+	}
+	return badEncodingFraudProof.Marshal()
+}
+
+// UnmarshalBinary converts binary to BadEncodingProof
+func (p *BadEncodingProof) UnmarshalBinary(data []byte) error {
+	in := pb.BadEncoding{}
+	if err := in.Unmarshal(data); err != nil {
+		return err
+	}
+	befp := UnmarshalBEFP(&in)
+	*p = *befp
+
+	return nil
+}
+
+// UnmarshalBEFP converts given data to BadEncodingProof
+func UnmarshalBEFP(befp *pb.BadEncoding) *BadEncodingProof {
+	return &BadEncodingProof{
+		BlockHeight: befp.Height,
+		Shares:      ipld.ProtoToShare(befp.Shares),
+	}
+}
```

Comment:
nit: Is this really Unmarshall, if doesn't covert actual bytes to the struct? Do we need this to be exported?

Reply:
`ProtoToBEFP` ?

Reply:
Ok